EIP-2718: Typed Transaction Envelope + EIP-2930

lightclient opened the pull request:
Hi all, I'm working on implementing [EIP-2718: Typed Transaction Envelope](https://eips.ethereum.org/EIPS/eip-2718). As of [ACD 92](https://github.com/ethereum/pm/blob/master/All%20Core%20Devs%20Meetings/Meeting%2092.md#22-eip-2718) it is considered EFI. It is a dependency for the following:
* [YOLOv2](https://github.com/ethereum/eth1.0-specs/projects/1#column-10680418) testnet
* [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559)
* [EIP-2711](https://eips.ethereum.org/EIPS/eip-2711)
* [EIP-2930](https://eips.ethereum.org/EIPS/eip-2930)
* [EIP-2803](https://eips.ethereum.org/EIPS/eip-1559)

Given the amount of code that this PR will eventually update, it would help me a lot to get some feedback earlier on to make sure this is headed in a good direction. Please let me know if there is a better way to request feedback.

lightclient commented:
Sorry, I meant to open this as a draft PR. If someone with write access to this repo could update it to a draft, that would be great.

lightclient commented:
I'm planning to update this to use the original `Transaction` struct + interface as `txdata`.

holiman commented:
Thanks for starting to work on this. It's a bit hard to judge if the interfaces are 'right' without more concrete examples, so I was wondering if you'd also be interested in implementing the [2930](https://eips.ethereum.org/EIPS/eip-2930) transaction type within the same PR?
Note, you wouldn't need to implement the actual handling of access lists, but use the concrete new transaction type to see if the abstraction model you have chosen is sufficient to accommodate it?

I was thinking of picking it up myself, but it seems like a bad idea for me to implement something that you've already gotten a good start on... ?

holiman commented:
Regarding your concrete questions, I don't really have any good answers now, I think it will be easier to give feedback with more concrete types implemented.

lightclient commented:
>  I was wondering if you'd also be interested in implementing the 2930 transaction type within the same PR?
Yep, I would be happy to.

lightclient commented on core/types/transaction.go:
```diff
@@ -128,150 +89,140 @@ func isProtectedV(V *big.Int) bool {
+	}
+
+	var dec id
+	if err := json.Unmarshal(input, &dec); err != nil {
```
Any thoughts on hoisting the hash out the `inner` tx type? The alternative is to put `Hash` in all `inner` tx types

lightclient commented on core/types/gen_tx_json.go:
```diff
@@ -11,10 +11,8 @@ import (
-// MarshalJSON marshals as JSON.
-func (t txdata) MarshalJSON() ([]byte, error) {
+// MarshalJSONWithHash marshals as JSON with a hash.
+func (t *LegacyTransaction) MarshalJSONWithHash(hash *common.Hash) ([]byte, error) {
```
Any thoughts on this modification to the `inner` tx marshaller?

fjl commented on core/types/legacy_tx.go:
```diff
@@ -0,0 +1,101 @@
+	S *big.Int `json:"s" gencodec:"required"`
+}
+
+func NewLegacyTransaction(nonce uint64, to common.Address, amount *big.Int, gasLimit uint64, gasPrice *big.Int, data []byte) *Transaction {
```
Most changes in this PR are due to the renaming of Transaction to LegacyTransaction.
I think it would be better to keep the `NewTransaction` constructor name to avoid the changes.

lightclient replied:
Makes sense, will revert.

lightclient commented:
@holiman @fjl, do you prefer the tests for this PR to be written natively in geth, as part of retesteth, or both?

holiman commented:
>  do you prefer the tests for this PR to be written natively in geth, as part of retesteth, or both?

We always want 'native' tests in the codebase. If there are other more high-level tests that's good too, but we don't want to rely on those only

lightclient commented on core/state_processor.go:
```diff
@@ -86,6 +86,10 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 // for the transaction, gas used and an error if the transaction failed,
 // indicating the block was invalid.
 func ApplyTransaction(config *params.ChainConfig, bc ChainContext, author *common.Address, gp *GasPool, statedb *state.StateDB, header *types.Header, tx *types.Transaction, usedGas *uint64, cfg vm.Config) (*types.Receipt, error) {
+	if !config.IsYoloV1(header.Number) && tx.Type() != types.LegacyTxId {
```
Curious is this is the appropriate place to make this consensus check?

fjl replied:
I think we should perform this check in the signer instead.

lightclient commented:
Hi @holiman, I implemented EIP-2930 using the EIP-2718 envelope. This PR is rebased against your EIP-2929 PR. For a simpler diff, I have [a PR](https://github.com/lightclient/go-ethereum/pull/3) on my own repo against your branch.

holiman commented:
I rebased this on top of `access_lists`. 
I think it'll be easier to review this once that one is merged

lightclient commented on core/types/transaction_signing.go:
```diff
@@ -102,6 +104,80 @@ type Signer interface {
 
+// HomesteadTransaction implements TransactionInterface using the
+// homestead rules.
+type YoloSigner struct{ chainId, chainIdMul *big.Int }
```
I updated the PR to wrap the EIP155Signer. Unfortunately, was only able to reuse the `Equal` function. You'll see that `Sender` and `Signature` functions need to only normalize `v` in the case of a legacy tx. Then `Hash` is reimplemented to support the access list tx.

lightclient commented:
Hi all, I've moved this out of the draft mode. Now that I've rebased against master with @holiman's EIP-2929 implementation and spent some time this week further cleaning the code up, It would be helpful to get a review and some guidance on how to best move forward. 

The current status of the PR:
* Implements EIP-2718 and EIP-2930
* Blockchain test ensuring proper gas accounting for EIP-2930 access list
* Serialization tests for txs + blocks
* Local two node test, verifying that RPC accepts EIP-2930 txs, mines them, and that another node can full sync it
* Currently syncing mainnet with #ac32793

Things that may be missing:
* New wire protocol -- this was discussed in a prior call by @fjl (note that EIP-2718 txs are currently working over eth/65 with this PR)
* Wire tests
* RPC tests
* update gas estimator?
* mempool tests?

Happy to iterate on the feedback. Thanks!

MariusVanDerWijden commented on core/types/transaction.go:
```diff
@@ -55,31 +56,24 @@ type Transaction struct {
-
 	// AccessList returns the transactions optional EIP-2930 access list.
 	AccessList() *AccessList
-
```
hey @lightclient could you please re-lint this file? It will be accepted as is, but the default linter will add some whitespaces here which means the next one to modify this file will have a bunch of changes that make git blaming harder

lightclient replied:
Hi @MariusVanDerWijden, I've rerun the linter (`make lint`), but am not seeing any changes? Is that the correct way to re-lint?

fjl replied:
I think what he meant is: please run `gofmt -w -s core/types/transaction.go`

lightclient replied:
Hmm after running, I'm still not seeing any changes. Is there something specific you are concerned with that I should modify?

lightclient commented:
![2020-10-30-1604067634_syncing](https://user-images.githubusercontent.com/14004106/97716244-00ba8e80-1a89-11eb-81af-ac6dfe1b9352.jpg)
Synced past 5.5m blocks so far.

rakita commented:
Hi @lightclient, I am implementing this in OpenEthereum and wanted to sync. The main question is how do you handle a list of transactions from for example [Transaction package](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#transactions-0x02).
This is RLP list that goes like: `rlpList[ rlpList(tx), rlpList(tx), ...)]`
if we strictly make a change as specified at 2718 this will become: `rlpList[ Type|tx, rlpList(tx), ...]` for this example i will need to change how to iterate over rlp list because now our Typed transaction does not have size encoded in rlp that is needed to iterate list, but size needs to be derived of type to know next item. Not sure how iterations are done in go rlp lib.

And second small thing to say, it is not specified in EIP-2930 , but I am okay to put chain_id in second place when calculating hash.

holiman commented on core/types/access_list_tx.go:
```diff
@@ -0,0 +1,124 @@
+	Recipient    *common.Address `json:"to"         rlp:"nil"` // nil means contract creation
+	Amount       *big.Int        `json:"value"      gencodec:"required"`
+	Payload      []byte          `json:"input"      gencodec:"required"`
+	Accesses     *AccessList     `json:"accessList" rlp:"nil"`
```
I don't quite understand why this is marked `rlp:"nil"` ?

holiman commented on core/state_processor.go:
```diff
@@ -86,6 +86,9 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 // for the transaction, gas used and an error if the transaction failed,
 // indicating the block was invalid.
 func ApplyTransaction(config *params.ChainConfig, bc ChainContext, author *common.Address, gp *GasPool, statedb *state.StateDB, header *types.Header, tx *types.Transaction, usedGas *uint64, cfg vm.Config) (*types.Receipt, error) {
+	if !config.IsYoloV2(header.Number) && tx.Type() != types.LegacyTxId {
+		return nil, ErrTxTypeNotSupported
+	}
```
Other than rejecting non-legacy txs before yolov2 (or whatever we later call it), don't we also need to
- Verify that the type is exactly the AccessList-type, if access lists are provided? (so we reject legacy-txs with access-lists, and foobar-typed txs in general)

Or is that done somewhere else? This cannot rely on txpool verifications, since it may come in the form of a maliciously mined block

lightclient replied:
Whenever any RLP bytes are decoded, it will determine if the transaction type is "known": https://github.com/ethereum/go-ethereum/pull/21502/files#diff-8fb609018eccff10931645347515b8ecc194992cc4ef9e84ebdea87682dd50c3R123

So it would be impossible for any other transaction type to be accepted as it will always error out. If you would prefer explicit checks in the state processor (or elsewhere?), I can add those. The legacy transaction cannot have an access list, because it's return is value is constant: https://github.com/ethereum/go-ethereum/pull/21502/files#diff-366ae3d3949ab2942dd8810a703bf0531f5ade9b7aa1d144e60aac3c2a7cf689R78

holiman commented on core/types/transaction_signing.go:
```diff
@@ -102,6 +104,78 @@ type Signer interface {
+// Sender returns the recovered addressed from a transaction's signature.
+// It assumes V does not store the chain id, unless the tx is of legacy type.
+func (s EIP2718Signer) Sender(tx *Transaction) (common.Address, error) {
+	if !tx.Protected() {
+		return HomesteadSigner{}.Sender(tx)
+	}
```
Is there an implicit assumption here that non-legacy transactions are protected? (if so: I don't think that assumption should be made)

holiman replied:
Unless it's clearly specified in the 2718-EIP that new-style transactions are `Protected`

lightclient replied:
I reordered the function so it is clear that there is no assumption as to whether typed txs are protected.

holiman commented on params/protocol_params.go:
```diff
@@ -60,20 +60,22 @@ const (
+	Create2Gas                uint64 = 32000 // Once per CREATE2 operation
+	SelfdestructRefundGas     uint64 = 24000 // Refunded following a selfdestruct operation.
+	MemoryGas                 uint64 = 3     // Times the address of the (highest referenced byte in memory + 1). NOTE: referencing happens on read, write and in instructions such as RETURN and CALL.
+	TxDataNonZeroGasFrontier  uint64 = 68    // Per byte of data attached to a transaction that is not equal to zero. NOTE: Not payable on data of calls between transactions.
```
You could leave a linebreak before the `Tx...` constants, imo

holiman reviewed (commented):
I think this looks pretty good

lightclient commented:
@rakita unfortunately we decided on the type concatenation after quite a bit of discussion, so I'm afraid it may need to be implemented as is. @holiman please let us know if you feel differently.

Good catch on the chain id. @holiman, do you think a change to EIP-2930 to move the chain id from `v` into its own field is acceptable? Now that we have a better way to differentiate between transaction types, this may be a welcome change for developers. I've written the PR in a way that does not allow for a transaction to be valid on multiple chains at once, but we could modify this behaviour if desired.

holiman commented:
> Edit: I see your comment now re: protected transactions. I fix this. Please let me know if I should keep or remove the chain id field. If we keep it, we'll need to decide upon a value that means a tx is not protected.

Hm -- I don't think it's something that should be decided _on this PR_. This PR should implement the EIP(s), and the EIP(s) should clarify exactly the formats of transactions, otherwise they are underspecified. Unless there's something I haven't understood.

holiman commented on core/types/access_list_tx.go:
```diff
@@ -97,7 +97,7 @@ func newAccessListTransaction(chainId *big.Int, nonce uint64, to *common.Address
 
 func (tx *AccessListTransaction) ChainId() *big.Int       { return tx.Chain }
-func (tx *AccessListTransaction) Protected() bool         { return true }
+func (tx *AccessListTransaction) Protected() bool         { return tx.Chain.Cmp(common.Big0) == 0 }
```
oldie but goldie: you can use `tx.Chain.BitLen() == 0` instead

holiman replied:
But shouldn't it be the other way around (both in the current code and my example?) `!= 0` ?

lightclient replied:
Ah, good catch. I had fixed this locally, but managed to not save the changes before committing. I'll update to `tx.Chain.BitLen() != 0` instead of using `Cmp`.

lightclient commented:
![2020-11-04-1604526835_2718_syncing](https://user-images.githubusercontent.com/14004106/98172020-e24ffb00-1ead-11eb-9838-23182bd68636.jpg)
Synced over 10mm blocks.

holiman commented on core/types/access_list_tx.go:
```diff
@@ -97,7 +97,7 @@ func newAccessListTransaction(chainId *big.Int, nonce uint64, to *common.Address
 
 func (tx *AccessListTransaction) ChainId() *big.Int       { return tx.Chain }
-func (tx *AccessListTransaction) Protected() bool         { return tx.Chain.Cmp(common.Big0) == 0 }
+func (tx *AccessListTransaction) Protected() bool         { return tx.Chain.BitLen() != 0 }
```
TODO: change to `true` instead

lightclient commented:
I've updated this PR to reflect the latest changes of EIP-2930.

holiman commented:
I modified one of the testcases a bit: https://github.com/holiman/go-ethereum/commit/e22dc637d6265f359eca5af43104847ceffe767a . I didn't push to your branch, because the test is now failing: 
```
    TestTransactionCoding: transaction_test.go:353: json decoding failed: missing required field 'accessList' for txdata
```
It happens when using `rawTx = NewAccessListTransaction(,.,.,. nil)`, so the accesslist is `nil`. And it's required by the json marshaller. So, basically

- If we want to ensure that it's always an empty list, and not `nil`, then we should reject it earlier, 
- Otherwise, if we're ok with `nil`, the json marshaller needs to be fixed. 

You can cherry-pick that commit if you want to fix it one way or the other

lightclient commented:
@holiman thanks for that. It seems best to enforce that json access list txs must include an access list. I've included your changes and updated the `newAccessListTransaction` function to initialize the tx's access list to an empty list in the event that `nil` is passed to the function.

holiman commented:
Rihgt now, you have defined 
```
type AccessList []AccessTuple
...
type AccessListTransaction struct {
...
	Amount       *big.Int        `json:"value"      gencodec:"required"`
	Payload      []byte          `json:"input"      gencodec:"required"`
	Accesses     *AccessList     `json:"accessList" gencodec:"required"`

	// Signature values
```
So `AccessListTransaction.Accesses` is essentially a `*[]AccessTuple`. Is it really a good idea to have it as a (potentially nil) pointer? Instead of having it be a slice `Accesses     AccessList     `json:"accessList" gencodec:"required"``?

holiman commented on core/state_transition.go:
```diff
@@ -104,10 +106,10 @@ func (result *ExecutionResult) Revert() []byte {
 
 // IntrinsicGas computes the 'intrinsic gas' for a message with the given data.
-func IntrinsicGas(data []byte, contractCreation, isHomestead bool, isEIP2028 bool) (uint64, error) {
+func IntrinsicGas(data []byte, accessList *types.AccessList, isContractCreation bool, isHomestead, isEIP2028, isEIP2718 bool) (uint64, error) {
```
I'm a bit torn here. Does it really make sense to have `isEIP2718`? If this method instead takes a (potentially nil) accessList, then we simply charge for it. 
As it is, if we ask about the intrinsic gas for an accesslist-type tx but say that it's not 2718, then this method will return the wrong intrinsic gas. 
If we skip `isEIP2718`, then at least the method cannot return values that doesn't make sense

lightclient replied:
That's a fair point. It's also poor wording, `is2718` is more akin to `isHomestead`. But I think you're right, it's okay to simply use the `accessList` value.

holiman commented:
I think something like this will be needed, unless I'm missing something, to prevent invalid transactions from being accepted by the txpool 

```diff 

diff --git a/core/tx_pool.go b/core/tx_pool.go
index 14785f846a..13922ca5e9 100644
--- a/core/tx_pool.go
+++ b/core/tx_pool.go
@@ -534,6 +534,12 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
 	if pool.currentMaxGas < tx.Gas() {
 		return ErrGasLimit
 	}
+	if tx.Type() != types.LegacyTxId && !pool.yoloV2 {
+		return ErrTxTypeNotSupported
+	}
+	if pool.yoloV2 && (tx.Type() != types.LegacyTxId && tx.Type() != types.AccessListTxId) {
+		return ErrTxTypeNotSupported
+	}
 	// Make sure the transaction is signed properly
 	from, err := types.Sender(pool.signer, tx)
 	if err != nil {
```

lightclient commented:
> So AccessListTransaction.Accesses is essentially a *[]AccessTuple. Is it really a good idea to have it as a (potentially nil) pointer? Instead of having it be a slice?
The main reason I created the additional type is for the convenience function `StorageKeys`. It seems verbose to sum the storage keys outside the transactions code. IMO, the main concern is that currently the convenience functions assume the caller has checked that `AccessList` is not `nil`. Would adding a `nil` check in those functions alleviate your concerns about using a pointer? Or do you still prefer to fallback to the tuple slice?

holiman commented:
As far as I understood the discussions on friday's ACD, and the discussions on discord afterwards. 

- The hash format of new transactions should be changed from `hash( [ 1, [ ...]])`, into `hash[1, [ ...] ) `
- We should have an additional wrapped legacy tx, which is `hash( 0, [...])` (?)
- The `DeriveSha` method needs to be extended to handle this aswell (perhaps splitting into `deriveShaLegacy` and `deriveSha2718`)

I may have gotten something bass-ackwards here, just a starting point to determine what changes are needed going forward.

lightclient commented:
> The DeriveSha method needs to be extended to handle this as well (perhaps splitting into deriveShaLegacy and deriveSha2718)

As written, I believe that `DeriveSha` will calculate the root correctly. This is because `EncodeRLP` is actually hacked in typed transactions to return the non-conforming RLP bytes `type || tx RLP encoding`. While this makes some things easier, it is concerning that `EncodeRLP` does *not* return conformant RLP. If we decide that `EncodeRLP` should return only the transaction's RLP encoding and the `type` would need to be assembled elsewhere, I'm not sure what the cleanest way to achieve that would be. In terms of encoding/decoding transactions themselves, it wouldn't be too bad as we could swap `decodeRlp`/`encodeRlp` with `decodeTx`/`encodeTx`, but I'm not sure how to best do this in recursive situations, such as block encoding/decoding.

@karalabe @fjl @holiman: do you mind weighing in on this?

holiman commented:
We discussed this today. So as long as `EncodeRLP`/`DecodeRLP` are internally consistent, then that's fine. 

However, this might be a tad moot, depending on how the continued discussion about SSZ goes.

rakita commented on core/types/block_test.go:
```diff
@@ -59,6 +59,53 @@ func TestBlockEncoding(t *testing.T) {
+}
+
+func TestEIP2718BlockEncoding(t *testing.T) {
+	blockEnc := common.FromHex("f90316f90211a00000000000000000000000000000000000000000000000000000000000000000a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347948888f1f195afa192cfee860698584c030f4c9db1a0ef1552a40b7165c3cd773806b9e0c165b75356e0314bf0706f279c729f51e017a08dae65c92dceddaa2fabec652b550258568abed0de8ae7482e662afa5cd6bef7a0cafe75574d59780665a97fbfd11365c7545aa8f1abf4e5e12e ...
```
Wanted to steal test data :) but I found something unexpected for the AccessList transaction

![Screenshot_5](https://user-images.githubusercontent.com/13179220/100992192-2d643900-3554-11eb-836c-32a7316391f9.png)

In OE i implemented it as `rlpBytes( 01 || rlpList([..]))` and confirmed it [here](https://ethereum-magicians.org/t/eip-2718-typed-transaction-envelope/4355/59?u=rakita) that this is okay, but in test data it is represented as `rlpList([01,...])`

lightclient replied:
Thanks for pointing that out @rakita, I will update it.

holiman commented:
@lightclient are you still working on getting this up to date with the latest spec updates?

lightclient commented:
@holiman yes, I'll have it up-to-date tomorrow.

holiman commented:
@lightclient are you still working on this?

lightclient commented:
@holiman yes I am still working on it. I'm sorry for the delays, I didn't make as much progress over the holidays as expected.

lightclient commented:
Getting off for now, but this last commit should pass the test suite. With that, I believe it is now up to spec and ready to be reviewed. I will go through once more tomorrow to look at it with fresher eyes to make sure I'm not missing something.

holiman commented:
I added some code to generate files for cross client testing, they might be removed later, but it's nice to be able to correlate these things early on.

lightclient commented on core/types/block.go:
```diff
@@ -169,7 +179,7 @@ type Body struct {
 	header       *Header
 	uncles       []*Header
-	transactions Transactions
+	transactions []*Transaction
```
Oops, I changed the `extblock` to use `Transactions` instead of `[]*Transaction` in an earlier version to hijack `EncodeRLP`, but when I manually reverted I accidentally also changed this one. I will undo this.

lightclient commented on core/types/block.go:
```diff
@@ -147,6 +147,16 @@ func rlpHash(x interface{}) (h common.Hash) {
 	return h
 }
 
+func typedRlpHash(typ uint8, x interface{}) (h common.Hash) {
```
Any thoughts on this approach to concatenating the type byte at the front of the RLP data?

DGKSK8LIFE reviewed (approved):
lgtm. Good typecheck.

holiman commented on core/types/receipt.go:
```diff
@@ -127,20 +133,36 @@ func NewReceipt(root []byte, failed bool, cumulativeGasUsed uint64) *Receipt {
 // DecodeRLP implements rlp.Decoder, and loads the consensus fields of a receipt
 // from an RLP stream.
 func (r *Receipt) DecodeRLP(s *rlp.Stream) error {
+	typ := uint64(LegacyTxId)
+	k, _, err := s.Kind()
+	if err != nil {
+		return err
+	}
+	// If the receipt isn't a list, it's likely typed - so pop off the first byte.
+	if k != rlp.List {
+		if typ, err = s.Uint(); err != nil {
+			return err
+		}
+	}
```
@fjl you're the codec-guy, does this look ok to you?

holiman commented:
I added another commit which adds support for creating transactions from the console (i.e also via RPC). While doing that, I also found some quirks in the signing portion, which were fixed. 

With the last commit, one can now do 

```
[user@work go-ethereum]$ ./build/bin/geth --dev console 
...
INFO [01-12|14:32:58.884] Initialised chain configuration          config="{ChainID: 1337 Homestead: 0 DAO: <nil> DAOSupport: false EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0 Petersburg: 0 Istanbul: 0, Muir Glacier: <nil>, YOLO v2: 0, Engine: clique}"
...
Welcome to the Geth JavaScript console!
...
> eth.fillTransaction({from: eth.accounts[0] , to: eth.accounts[0] , accessList: [{"address": "0x0000000000000000000000000000000000001337","storageKeys": ["0x0000000000000000000000000000000000000000000000000000000000000000"]}] })
{
  raw: "0xb85f01f85c82053980018262d4943cee8c66debd7a95a6f3fbd9ca40c530db7907338080f838f7940000000000000000000000000000000000001337e1a00000000000000000000000000000000000000000000000000000000000000000808080",
  tx: {
    accessList: [{
        address: "0x0000000000000000000000000000000000001337",
        storageKeys: [...]
    }],
    chainId: "0x539",
    gas: "0x62d4",
    gasPrice: "0x1",
    hash: "0xd7b1fbac95f4cf613da5611872b034407d3386c372c22cea0b2babe52c3c16e9",
    input: "0x",
    nonce: "0x0",
    r: "0x0",
    s: "0x0",
    to: "0x3cee8c66debd7a95a6f3fbd9ca40c530db790733",
    type: "0x1",
    v: "0x0",
    value: "0x0"
  }
}
> eth.sendTransaction({from: eth.accounts[0] , to: eth.accounts[0] , accessList: [{"address": "0x0000000000000000000000000000000000001337","storageKeys": ["0x0000000000000000000000000000000000000000000000000000000000000000"]}] })
INFO [01-12|14:33:26.285] Setting new local account                address=0x3CEe8c66DEbd7A95a6f3FbD9ca40C530Db790733
INFO [01-12|14:33:26.286] Submitted transaction                    fullhash=0x81b7d49d5ce39b63348998c0c5130a495b2a9b30dc7c6e2c2db6e7ca09279b72 recipient=0x3CEe8c66DEbd7A95a6f3FbD9ca40C530Db790733
"0x81b7d49d5ce39b63348998c0c5130a495b2a9b30dc7c6e2c2db6e7ca09279b72"
INFO [01-12|14:33:26.286] Commit new mining work                   number=1 sealhash="be292b…7f0e2a" uncles=0 txs=1 gas=25300 fees=2.53e-14 elapsed="869.903µs"
INFO [01-12|14:33:26.287] Successfully sealed new block            number=1 sealhash="be292b…7f0e2a" hash="83a527…fb0800" elapsed="390.182µs"
INFO [01-12|14:33:26.287] 🔨 mined potential block                  number=1 hash="83a527…fb0800"
INFO [01-12|14:33:26.290] Commit new mining work                   number=2 sealhash="794a75…574308" uncles=0 txs=0 gas=0     fees=0        elapsed=3.437ms
>
```

holiman commented on core/types/transaction.go:
```diff
@@ -25,191 +26,174 @@ import (
 	"github.com/ethereum/go-ethereum/rlp"
 )
 
-//go:generate gencodec -type txdata -field-override txdataMarshaling -out gen_tx_json.go
```
@lightclient you removed this gen-command, but you have also modified the `gen_tx_json.go` generated file. How did you generate that one? The command should be added to the definition-file, like here, so one can do e.g.  `go generate transaction.go ` to regenerate the files.

lightclient replied:
Oh, I see. I manually updated the `gen_tx_json.go`.

holiman replied:
I see... This is a bit unconventional, I'll see if I can make it work differently, by generating files for legacy / acl separately, and have a custom marshal/unmarshal for `Transaction` which tests them both

holiman replied:
I don't know. It seems pretty tricky to auto-generate the corresponding code for this, so maybe it's indeed better to have it the way you've implemented it. If so, we should change the name from `gen_tx.json` to `transaction_marshalling.go` or something. 

@fjl what do you think?

holiman commented on core/types/gen_tx_json.go:
```diff
@@ -11,91 +11,181 @@ import (
 // UnmarshalJSON unmarshals from JSON.
-func (t *txdata) UnmarshalJSON(input []byte) error {
+func (t *Transaction) UnmarshalJSON(input []byte) error {
+	type typ struct {
+		Type *hexutil.Uint64 `json:"type" rlp:"-"`
+	}
```
Why did this wind up in it's own struct? Being of pointer type, seems like it could be read simultaneously as everything else, regardless of whether it's present or not?

lightclient replied:
I'm not sure now, it looks like it would be fine to be in the `txdata` struct.

holiman replied:
Yeah, I didn't mention it, but I already changed that, it worked well

lightclient replied:
I noticed when I went to change it :)

MariusVanDerWijden commented on core/types/transaction.go:
```diff
@@ -25,191 +26,174 @@ import (
-		return v != 27 && v != 28
+		return v != 27 && v != 28 && v != 1 && v != 0
 	}
 	// anything not 27 or 28 is considered protected
```
stale comment

MariusVanDerWijden commented:
I started playing around a bit with access lists and the new tx format and have some suggestions regarding the API.

I think it would be good to have a way to convert the new `types.LegacyTransaction` and `types.AccessListTransaction` to the new `types.Transaction` format builtin otherwise it is not possible to convert a signed legacy transaction to a normal transaction and execute it, Don't know how big of an use case this is, since creating a legacy transaction with `types.NewTransaction()` already wraps the tx properly. 
Edit: now that I've thought about it a bit more, maybe it is not such a good idea ...
```
// Conversion from LegacyTransaction to Transaction drops the signature
tx *types.LegacyTransaction
exec := types.NewTransaction(tx.AccountNonce, *tx.To(), tx.Amount, tx.GasLimit, tx.GasPrice(), tx.Data())

// Maybe for people that would rather work with the raw types we could add something like this:

// Wrap converts a legacy transaction into a eip-2718 wrapped transaction
func (tx *LegacyTransaction) Wrap() *Transaction {
	return &Transaction{
		typ:   LegacyTxId,
		inner: &tx,
		time:  time.Now(),
	}
}
```

Additionally I think it is necessary to provide a way to get an access list from a transaction by executing it on some state either from the simulated backend or the ethclient. 
My idea would be something along the lines of this: (snippet doesn't work because `b.Blockchain().State()` creates a new statedb instead of returning the current one.
```go
func legacyTxToAccessListTx(b *backends.SimulatedBackend, tx *types.LegacyTransaction) (*types.Transaction, error) {
// ApplyTX
exec := types.NewTransaction(tx.AccountNonce, *tx.To(), tx.Amount, tx.GasLimit, tx.GasPrice(), tx.Data())
b.SendTransaction(context.Background(), exec)
// Query current statedb
state, err := b.Blockchain().State()
if state == nil || err != nil {
	return nil, err
}
// retrieve access list
acl := state.AccessList()
return types.NewAccessListTransaction(tx.ChainId(), tx.Nonce(), *tx.To(), tx.Amount, tx.GasLimit, tx.Price, tx.Data(), acl), nil
```

For that the statedb would need a way to provide the access lists:
This snippet converts the access list format used in the statedb to the format used in `types.AccessList`

```go
func (s *StateDB) AccessList() *types.AccessList {
	list := s.accessList.Copy()
	acl := make([]types.AccessTuple, 0, len(list.addresses))
	for addr, idx := range list.addresses {
		var tuple types.AccessTuple
		tuple.Address = &addr
		keys := make([]*common.Hash, 0, len(list.slots[idx]))
		for key, _ := range list.slots[idx] {
			tuple.StorageKeys = append(tuple.StorageKeys, &key)
		}
		tuple.StorageKeys = keys
		acl = append(acl, tuple)
	}
	cast := types.AccessList(acl)
	return &cast
}
```

Don't think that this has to be added to this pr necessarily, we could always add it in the future

holiman commented:
TODO @holiman

- Remove the not-very-good json blocktests again, 
- Rebase on master
- Squash it a bit
- Run as full-sync on benchmarkers + fast-sync on benchmarkers

holiman commented:
The full-syncs both balooned up on memory and died :/

holiman commented:
Actually, I'm not sure what happened. It seems both machines (aws instances) lost their network capabilities, and failed to send logs + metrics, but when the machine was rebooted, the geth process was still up, and managed to shutdown in time and restart gracefully.

holiman commented:
> Additionally I think it is necessary to provide a way to get an access list from a transaction by executing it on some state either from the simulated backend or the ethclient.

@MariusVanDerWijden yes, I agree, and picking it out from the statedb after-the-fact seems like  pretty good way to go about it. It needs to be 'cleansed' from `tx.origin`, `tx.to` and the precompiles, though, which are added 'for free' by the system.

And yes, I think it should be a separate PR.  Also, I don't think the functionality should be restricted to `legacyTx`. Consider the case where you do a first al-transaction. it fails, because at some point it's choked. You get the "updated" `al`, run it again, find another chokehole, add that accounts/slot, run it again, and done.

ryanschneider commented:
If I'm reading the code of `Transaction.Hash` (https://github.com/lightclient/go-ethereum/blob/67bcc691b929c36cd0ff18ecf8e972ee10692d4d/core/types/transaction.go#L188) correctly this is computing the hash of an EIP-2930 tx as `keccak256( rlp(1) || rlp([chainId, nonce, gasPrice, gasLimit, to, value, data, access_list]))` but the EIP states that the signature should be over `keccak256(rlp([1, chainId, nonce, gasPrice, gasLimit, to, value, data, access_list]))` is that an issue?

ryanschneider commented:
> If I'm reading the code of `Transaction.Hash` (https://github.com/lightclient/go-ethereum/blob/67bcc691b929c36cd0ff18ecf8e972ee10692d4d/core/types/transaction.go#L188) correctly this is computing the hash of an EIP-2930 tx as `keccak256( rlp(1) || rlp([chainId, nonce, gasPrice, gasLimit, to, value, data, access_list]))` but the EIP states that the signature should be over `keccak256(rlp([1, chainId, nonce, gasPrice, gasLimit, to, value, data, access_list]))` is that an issue?

Discussed this in Slack, it's not an issue just me misunderstanding the EIP, never mind. :)

holiman commented:
@lightclient now we have 
```
// EncodeRLP implements rlp.Encoder
// For legacy transactions, it outputs rlp(tx.inner). For typed transactions
// it outputs rlp(tx.type || tx.inner).
func (tx *Transaction) EncodeRLP(w io.Writer) error {
```
and 
```
// Raw transactions are used for internal processes which need the raw
// consensus representation of typed transactions, not the RLP string
// wrapped version (e.g. type || payload vs. rlp(type || payload)).
type rawtx Transaction

func (tx *rawtx) EncodeRLP(w io.Writer) error {
```
And at some times, we rlp-encode as a wrapped rlp-list, and at other times, as a prefixed opaque data structure. It would be good to make it clearer when we use what, and why.

fjl commented:
Added some changes to the RLP encoding and rebased on master branch.

fjl commented:
I think there a couple of minor changes still needed before it can be merged:

- The API of `Transaction` copies fields in accessor methods. This is to make the transaction object immutable. With the EIP-2718 changes, we now copy some fields two times because the methods of `innerTx` also perform this copy. 
- All signers except EIP2718Signer should reject non-legacy transactions. This is also the right way to implement the check for support of typed txs in package core: it calls `types.MakeSigner` for each block to find the applicable signer, then validates block txs with it. If EIP2718 is not activated yet, MakeSigner will return EIP155Signer and therefore reject typed txs.
- We can probably find a nicer API for the transaction constructor functions. The old `NewTransaction` / `NewContractCreation` way is really annoying to use in practice, we just never bothered to change it. How about something like `func NewTx(innerTx) *Transaction` where callers pass `LegacyTransaction` or `AccessListTransaction` and fill out the fields of the struct instead of passing 6+ positional parameters?

fjl commented:
Looks like I broke the StackTrie tests because they would like to define their own `DerivableList`. I think we should move those tests to core/types, will do it tomorrow.

holiman commented:
Current status, after recent changes. 
```
./build/bin/geth --dev console
...
> eth.fillTransaction({from: eth.accounts[0] , to: eth.accounts[0] , accessList: [{"address": "0x0000000000000000000000000000000000001337","storageKeys": ["0x0000000000000000000000000000000000000000000000000000000000000000"]}] })
{
  raw: "0xb85f01f85c82053980018262d4940993cb059574b88c78849139cb91f7d85f8f289a8080f838f7940000000000000000000000000000000000001337e1a00000000000000000000000000000000000000000000000000000000000000000808080",
  tx: {
    accessList: [{
        address: "0x0000000000000000000000000000000000001337",
        storageKeys: [...]
    }],
    chainId: "0x539",
    gas: "0x62d4",
    gasPrice: "0x1",
    hash: "0xb0b747850b660e1f809b5ce7ece72bac230b31aee6bc26ef102b0a8b1c2b5e7f",
    input: "0x",
    nonce: "0x0",
    r: "0x0",
    s: "0x0",
    to: "0x0993cb059574b88c78849139cb91f7d85f8f289a",
    type: "0x1",
    v: "0x0",
    value: "0x0"
  }
}
> eth.sendTransaction({from: eth.accounts[0] , to: eth.accounts[0] , accessList: [{"address": "0x0000000000000000000000000000000000001337","storageKeys": ["0x0000000000000000000000000000000000000000000000000000000000000000"]}] })
"0x9009f48b23a08bcfb91e7bf262fe9c536ccde626a631a6b65316697032054672"
> eth.getTransaction("0x9009f48b23a08bcfb91e7bf262fe9c536ccde626a631a6b65316697032054672")
{
  accessList: [{
      address: "0x0000000000000000000000000000000000001337",
      storageKeys: ["0x0000000000000000000000000000000000000000000000000000000000000000"]
  }],
  blockHash: "0xd7b588d1b264429d693d0a4405e913930f22cf5e1cb87c9956c5c0203a7ec194",
  blockNumber: 1,
  chainId: "0x539",
  from: "0x0993cb059574b88c78849139cb91f7d85f8f289a",
  gas: 25300,
  gasPrice: 1,
  hash: "0x9009f48b23a08bcfb91e7bf262fe9c536ccde626a631a6b65316697032054672",
  input: "0x",
  nonce: 0,
  r: "0xf496c34c3e9a838a459b3d42869efb9b79d020eccefc68ab55a4bfe6044a271e",
  s: "0x7a7df18d7c6b8f7981ec8f3360812e038e639d6e752302209c3d3593e857b2ed",
  to: "0x0993cb059574b88c78849139cb91f7d85f8f289a",
  transactionIndex: 0,
  type: "0x1",
  v: "0x0",
  value: 0
}
> eth.getRawTransaction("0x9009f48b23a08bcfb91e7bf262fe9c536ccde626a631a6b65316697032054672")
"0xb89f01f89c82053980018262d4940993cb059574b88c78849139cb91f7d85f8f289a8080f838f7940000000000000000000000000000000000001337e1a0000000000000000000000000000000000000000000000000000000000000000080a0f496c34c3e9a838a459b3d42869efb9b79d020eccefc68ab55a4bfe6044a271ea07a7df18d7c6b8f7981ec8f3360812e038e639d6e752302209c3d3593e857b2ed"
> 
```

As far as I understand things, the `getRawTransaction` (and, by extension, `fillTransaction`) should return the un-prefixed form. 

cc @ryanschneider-- are there any other discrepancies that you can see?

holiman commented:
Hm, it's not even agreeing with itself
```
> eth.fillTransaction({from: eth.accounts[0] , to: eth.accounts[0] , accessList: [{"address": "0x0000000000000000000000000000000000001337","storageKeys": ["0x0000000000000000000000000000000000000000000000000000000000000001"]}] })
{
  raw: "0xb85f01f85c82053901018262d4940993cb059574b88c78849139cb91f7d85f8f289a8080f838f7940000000000000000000000000000000000001337e1a00000000000000000000000000000000000000000000000000000000000000001808080",
  tx: {
    accessList: [{
        address: "0x0000000000000000000000000000000000001337",
        storageKeys: [...]
    }],
    chainId: "0x539",
    gas: "0x62d4",
    gasPrice: "0x1",
    hash: "0xddf44e2c40c6186b253b0c89bd5b5d6afdbe69e2cf02cd92cf569706a3a9143e",
    input: "0x",
    nonce: "0x1",
    r: "0x0",
    s: "0x0",
    to: "0x0993cb059574b88c78849139cb91f7d85f8f289a",
    type: "0x1",
    v: "0x0",
    value: "0x0"
  }
}
> eth.sendRawTransaction("0xb85f01f85c82053901018262d4940993cb059574b88c78849139cb91f7d85f8f289a8080f838f7940000000000000000000000000000000000001337e1a00000000000000000000000000000000000000000000000000000000000000001808080")
WARN [02-12|11:58:21.620] Served eth_sendRawTransaction            reqid=20 t="139.407µs" err="invalid sender"
Error: invalid sender
	at web3.js:6347:37(47)
	at web3.js:5081:62(37)
	at <eval>:1:23(4)

> eth.sendRawTransaction("0x01f85c82053901018262d4940993cb059574b88c78849139cb91f7d85f8f289a8080f838f7940000000000000000000000000000000000001337e1a00000000000000000000000000000000000000000000000000000000000000001808080")
WARN [02-12|11:58:35.002] Served eth_sendRawTransaction            reqid=21 t="81.455µs"  err="rlp: expected List"
Error: rlp: expected List
	at web3.js:6347:37(47)
	at web3.js:5081:62(37)
	at <eval>:1:23(4)
```

holiman commented:
@fjl I'm thinking something along these lines: 
```diff
diff --git a/internal/ethapi/api.go b/internal/ethapi/api.go
index 74ef76c915..060cebdd5e 100644
--- a/internal/ethapi/api.go
+++ b/internal/ethapi/api.go
@@ -1648,7 +1648,7 @@ func (s *PublicTransactionPoolAPI) FillTransaction(ctx context.Context, args Sen
 	}
 	// Assemble the transaction and obtain rlp
 	tx := args.toTransaction()
-	data, err := rlp.EncodeToBytes(tx)
+	data, err := tx.EncodeBinary()
 	if err != nil {
 		return nil, err
 	}
@@ -1658,8 +1658,8 @@ func (s *PublicTransactionPoolAPI) FillTransaction(ctx context.Context, args Sen
 // SendRawTransaction will add the signed transaction to the transaction pool.
 // The sender is responsible for signing the transaction and using the correct nonce.
 func (s *PublicTransactionPoolAPI) SendRawTransaction(ctx context.Context, encodedTx hexutil.Bytes) (common.Hash, error) {
-	tx := new(types.Transaction)
-	if err := rlp.DecodeBytes(encodedTx, tx); err != nil {
+	tx, err := new(types.Transaction).DecodeBinary(encodedTx)
+	if err != nil {
 		return common.Hash{}, err
 	}
 	return SubmitTransaction(ctx, s.b, tx)

```

ryanschneider commented:
@holiman I'll try to take a look today, it looks like you all are running into some of the same issues I am trying to update my transaction parsing code, which previously assumed that transactions were always RLP encoded, but I believe post-2718 that assumption is no longer valid, since `0x01f85c82053901018262d4940993...` isn't valid RLP anymore (since it's the concatenation of `0x01` and an RLP list).

IMO if you have any transaction encoding code returning or accepting RLP types versus `[]byte` for transactions post-2718 it's probably a code smell that something still needs refactored, I think your latest comment addresses that.

fjl commented on core/state_transition.go:
```diff
@@ -105,10 +107,10 @@ func (result *ExecutionResult) Revert() []byte {
 
 // IntrinsicGas computes the 'intrinsic gas' for a message with the given data.
-func IntrinsicGas(data []byte, contractCreation, isHomestead bool, isEIP2028 bool) (uint64, error) {
+func IntrinsicGas(data []byte, accessList *types.AccessList, isContractCreation bool, isHomestead, isEIP2028 bool) (uint64, error) {
```
Should we just pass `params.Rules` to this function instead of passing `isXXX` for each fork? I think this function will only ever get messier, might be a good time to prevent that.

holiman replied:
It's used from the txpool also, which doesn't have params.Rules per se

fjl replied:
Yes, but we can easily make it available in txpool. We can create `Rules` from `ChainConfig` and block number on every block new and use it to validate. Passing `types.Message` won't work for that though, we don't want to call `AsMessage` on every tx just for this.

fjl commented:
Some LES tests were broken by the update to receipt encoding. I think this is because LES expects the hash of downloaded receipts to match the hash of their RLP encoding, which is not true for typed receipts. Need to investigate this more.

fjl commented:
LES receipt response validation is actually pretty curious: https://github.com/ethereum/go-ethereum/blob/7d1b711c7d0f27efd7772c81bb73b9b29720515a/les/odr_requests.go#L177

No idea why this works at all right now.

holiman commented:
> LES receipt response validation is actually pretty curious, it does DeriveSha on a single receipt and expects it to match the block header?

The `receipts` is a slice of `Receipts` so `receipt` is a `Receipts`. Fun.

holiman commented:
@fjl I think the receipts is borked. Doing `./build/bin/geth --yolov3 --datadir /tmp/yolo3-snap --verbosity 4 --syncmode=snap` seems to download mostly everything fine, but it get's stuck in 
```
DEBUG[02-14|14:27:27.697] Fetching batch of receipts               id=a6c693cb158d5ea2 conn=dyndial count=2
DEBUG[02-14|14:27:27.811] Fetching batch of receipts               id=a6c693cb158d5ea2 conn=dyndial count=2
DEBUG[02-14|14:27:27.931] Fetching batch of receipts               id=a6c693cb158d5ea2 conn=dyndial count=2
DEBUG[02-14|14:27:28.045] Fetching batch of receipts               id=a6c693cb158d5ea2 conn=dyndial count=2
DEBUG[02-14|14:27:28.163] Fetching batch of receipts               id=a6c693cb158d5ea2 conn=dyndial count=2
DEBUG[02-14|14:27:28.272] Fetching batch of receipts               id=a6c693cb158d5ea2 conn=dyndial count=2
DEBUG[02-14|14:27:28.380] Fetching batch of receipts               id=a6c693cb158d5ea2 conn=dyndial count=2

```

holiman commented:
For some reason, `fast-sync` fails too. Relevant logs
```
DEBUG[02-14|14:30:59.590] Fetching batch of state data             id=a6c693cb158d5ea2 conn=dyndial count=1
DEBUG[02-14|14:30:59.590] Directing header downloads               peer=a6c693cb origin=1
...
TRACE[02-14|14:30:59.701] Received node data response              peer=a6c693cb158d5ea297eb69f6bedff152e495f318024ab826e74b9a8d178bf6d4 count=0   dropped=false timeout=false
WARN [02-14|14:30:59.701] Node data write error                    err="trie node b205a0…769afa failed with all peers (1 tries, 1 peers)"
INFO [02-14|14:30:59.701] Downloader queue stats                   receiptTasks=0 blockTasks=0 itemSize=0.00B throttle=8192
DEBUG[02-14|14:30:59.701] Header download terminated               peer=a6c693cb
```
The latest stateroot is `0xb205a05b6fbe95873b829eeea45f562d35c825a128234191f40e388773769afa` (it's been the same root for quite a while, being clique)

ryanschneider commented:
Not sure if this is related by the `from` in `eth_getTransactionReceipt` seems to be the zero address:

```
> eth.getTransactionReceipt("0x058c2fc78309580cd00b27baf88d6dc4b63dbff41c2f916b1b5126450e9e04ec")
{
  blockHash: "0x966e7b0d0ccd1cf28c2415f48fec456d66019a4b7065d8cac1af957f451b25eb",
  blockNumber: 70,
  contractAddress: null,
  cumulativeGasUsed: 39600,
  from: "0x0000000000000000000000000000000000000000",
  gasUsed: 39600,
  logs: [],
  logsBloom: "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  status: "0x1",
  to: "0x8a8eafb1cf62bfbeb1741769dae1a9dd47996192",
  transactionHash: "0x058c2fc78309580cd00b27baf88d6dc4b63dbff41c2f916b1b5126450e9e04ec",
  transactionIndex: 0
}
> eth.getTransaction("0x058c2fc78309580cd00b27baf88d6dc4b63dbff41c2f916b1b5126450e9e04ec")
{
  accessList: [{
      address: "0x0000000000000000000000000000000000001337",
      storageKeys: ["0x0000000000000000000000000000000000000000000000000000000000000000", "0x0000000000000000000000000000000000000133700000000000000000000000"]
  }, {
      address: "0x0000000000000000000000000000000000001337",
      storageKeys: ["0x0000000000000000000000000000000000000000000000000000000000000000", "0x0000000000000000000000000000000000000133700000000000000000000000"]
  }, {
      address: "0x0000000000000000000000000000000000001337",
      storageKeys: ["0x0000000000000000000000000000000000000000000000000000000000000000", "0x0000000000000000000000000000000000000133700000000000000000000000"]
  }],
  blockHash: "0x966e7b0d0ccd1cf28c2415f48fec456d66019a4b7065d8cac1af957f451b25eb",
  blockNumber: 70,
  chainId: "0x796f6c6f763378",
  from: "0x8a8eafb1cf62bfbeb1741769dae1a9dd47996192",
  gas: 39600,
  gasPrice: 1000000000,
  hash: "0x058c2fc78309580cd00b27baf88d6dc4b63dbff41c2f916b1b5126450e9e04ec",
  input: "0x",
  nonce: 4,
  r: "0x73d7d5edb4e10bce390d3ccad5477a7e6f4a80bae62917c181c5b8c77053b840",
  s: "0x296a7c23f768fd7dfaae2af7af0d98268fe9a17a8f582dbcacce1e4c0c999fe8",
  to: "0x8a8eafb1cf62bfbeb1741769dae1a9dd47996192",
  transactionIndex: 0,
  type: "0x1",
  v: "0x0",
  value: 0
}
```

winsvega commented:
The access list tests are generated on this PR

https://github.com/ethereum/tests/pull/774

ryanschneider commented:
BTW the receipt `from` field looks fine on legacy txs, for example `0x7d056b71d0860c0c5ad9beb7894d204b871e5b9153538b13841f15606c051620` (on the same account as the one with the sender bug).

fjl commented on internal/ethapi/api.go:
```diff
@@ -1427,10 +1427,18 @@ func (s *PublicTransactionPoolAPI) GetTransactionReceipt(ctx context.Context, ha
 	var signer types.Signer = types.FrontierSigner{}
 	if tx.Protected() {
-		signer = types.NewEIP155Signer(tx.ChainId())
+		switch txType {
```
This is not correct. We should use `types.MakeSigner` with the block number here. I'll fix it.

ryanschneider replied:
Cool, I took a stab at a fix here: https://github.com/ryanschneider/go-ethereum/commit/1ce3baeb2f947dea48719d7a206f748574c471d5  @fjl feel free to cherry-pick in if it looks correct.  Note that if this fix is correct we probably need to do something similar in `newRPCTransaction`.

fjl replied:
I fixed it in an even-simpler way. Still thinking about the right solution for newRPCTransaction

holiman commented on core/types/transaction_signing.go:
```diff
@@ -302,9 +318,14 @@ func (fs FrontierSigner) Hash(tx *Transaction) common.Hash {
-func (fs FrontierSigner) Sender(tx *Transaction) (common.Address, error) {
-	v, r, s := tx.RawSignatureValues()
-	return recoverPlain(fs.Hash(tx), r, s, v, false)
+func decodeSignature(sig []byte) (r, s, v *big.Int) {
+	if len(sig) != crypto.SignatureLength {
+		panic(fmt.Sprintf("wrong size for signature: got %d, want %d", len(sig), crypto.SignatureLength))
+	}
```
How come you chose `panic` here? The callers (`SignatureValues`) all return `error`, so they'd be able to handle if this method returned an error?

fjl replied:
The panic was there before.

holiman commented on core/types/transaction_signing.go:
```diff
@@ -126,27 +129,28 @@ func (s EIP2718Signer) Sender(tx *Transaction) (common.Address, error) {
-	}
 	switch tx.Type() {
 	case LegacyTxType:
+		R, S, V = decodeSignature(sig)
```
Wouldn't it be less copy-pastey if this just did `return s.EIP155Signer.SignatureValues(tx, sig)` ?

karalabe commented on core/types/access_list_tx.go:
```diff
@@ -0,0 +1,126 @@
+type AccessTuple struct {
+	Address     *common.Address `json:"address"        gencodec:"required"`
+	StorageKeys []*common.Hash  `json:"storageKeys"    gencodec:"required"`
+}
```
I don't really see a reason why these would be pointers internally. Unless I'm missing something, we should probably fix this.

karalabe replied:
Also please add type and method docs throughout

fjl replied:
Fixed

karalabe commented on core/tx_pool.go:
```diff
@@ -229,6 +229,7 @@ type TxPool struct {
 	mu          sync.RWMutex
 
 	istanbul bool // Fork indicator whether we are in the istanbul stage.
+	eip2718  bool // Fork indicator whether we are using EIp-2718 type transactions.
```
EIp -> EIP

karalabe commented on core/types/gen_receipt_json.go:
```diff
@@ -16,6 +16,7 @@ var _ = (*receiptMarshaling)(nil)
 // MarshalJSON marshals as JSON.
 func (r Receipt) MarshalJSON() ([]byte, error) {
 	type Receipt struct {
+		Type              uint8          `json:"type,omitempty"`
```
Just a mental note here, this will disable Types with id `0`. Just something to be aware if we ever want to wrap legacy txs with a zero.

karalabe commented on core/types/hashing.go:
```diff
@@ -0,0 +1,112 @@
+	New: func() interface{} { return new(bytes.Buffer) },
+}
+
+func rlpHash(x interface{}) (h common.Hash) {
```
I neeeed a method doc :)

karalabe commented on core/types/hashing_test.go:
```diff
@@ -0,0 +1,212 @@
+package types_test
```
Add copyright header pls

karalabe commented on core/types/legacy_tx.go:
```diff
@@ -0,0 +1,100 @@
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type LegacyTx struct {
```
docs pls

karalabe replied:
throughout

karalabe commented on core/types/receipt.go:
```diff
@@ -38,6 +38,8 @@ var (
 	receiptStatusSuccessfulRLP = []byte{0x01}
 )
 
+var errEmptyTypedReceipt = errors.New("empty typed receipt bytes")
```
Please add a doc that explains where this err is used.

karalabe commented on core/tx_pool.go:
```diff
@@ -535,7 +536,11 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
+	// Accept only legacy transactions if before 2718/2930.
+	if !pool.eip2718 && tx.Type() != types.LegacyTxType {
+		return ErrTxTypeNotSupported
+	}
```
Imho this check should be the first one. Don't even look any any of the fields for an unsupported tx type.

karalabe commented on core/types/transaction.go:
```diff
@@ -25,191 +26,238 @@ import (
-	return &Transaction{
-		data: d,
-		time: time.Now(),
+	// It's an EIP-2718 typed TX envelope.
```
Nitpick, Receipt encoding explicitly checks for AccessListTxType and rejects other types vs. this encoder accepts everything. I think we should add that check here too, since we don't know how the encoding will be for new tx types. The spec says it's an opaque blob, not necesarilly rlp.

karalabe commented on core/types/transaction.go:
```diff
@@ -25,191 +26,238 @@ import (
+	tx.typ = inner.Type()
+	tx.inner = inner
+	tx.time = time.Now()
+	if size > 0 {
```
Can the size be `0` to be worth this check?

ryanschneider commented:
^ 6a18663 broke GetTransactionReceipt, needs:

```diff
-	if txType != types.LegacyTxType {
+ 	if txType := tx.Type(); txType != types.LegacyTxType {
```

Here: https://github.com/ethereum/go-ethereum/blob/6a18663c677eca546c92482e74af822667c7f35d/internal/ethapi/api.go#L1458

holiman commented:
Some remaining nits
`internal/ethapi/api.go:L1798` should not directly call a signer this way. Probably it should use the pool signer?
```
		var signer types.Signer = types.HomesteadSigner{}
		if p.Protected() {
			signer = types.NewEIP155Signer(p.ChainId())
		}

```

`accounts/keystore/keystore.go`: check the type of the tx, and sign with 2718 if it's an access list tx?
```
func (ks *KeyStore) SignTxWithPassphrase(a accounts.Account, passphrase string, tx *types.Transaction, chainID *big.Int) (*types.Transaction, error) {
	_, key, err := ks.getDecryptedKey(a, passphrase)
	if err != nil {
		return nil, err
	}
	defer zeroKey(key.PrivateKey)

	// Depending on the presence of the chain ID, sign with EIP155 or homestead
	if chainID != nil {
		return types.SignTx(tx, types.NewEIP155Signer(chainID), key.PrivateKey)
	}
	return types.SignTx(tx, types.HomesteadSigner{}, key.PrivateKey)
}
```

graphql/graphql.go
```
func (t *Transaction) From(ctx context.Context, args BlockNumberArgs) (*Account, error) {
	tx, err := t.resolve(ctx)
	if err != nil || tx == nil {
		return nil, err
	}
	var signer types.Signer = types.HomesteadSigner{}
	if tx.Protected() {
		signer = types.NewEIP155Signer(tx.ChainId())
	}
	from, _ := types.Sender(signer, tx)

```

Should use `types.MakeSigner`? :
```
func (b *SimulatedBackend) SendTransaction(ctx context.Context, tx *types.Transaction) error {
	b.mu.Lock()
	defer b.mu.Unlock()

	sender, err := types.Sender(types.NewEIP155Signer(b.config.ChainID), tx)
```

Perhaps time to remove this: 
```
// Deprecated: use EthereumClient.TransactionSender
func (tx *Transaction) GetFrom(chainID *BigInt) (address *Address, _ error) {
```
The `light/txpool` needs an update: 
```
// NewTxPool creates a new light transaction pool
func NewTxPool(config *params.ChainConfig, chain *LightChain, relay TxRelayBackend) *TxPool {
	pool := &TxPool{
		config:      config,
		signer:      types.NewEIP155Signer(config.ChainID),

```
This check should probably be used in other places aswell, so maybe it can live somewhere so we don't have to copy-paste it everywhere: 
```
// newRPCTransaction returns a transaction that will serialize to the RPC
// representation, with the given location metadata set (if available).
func newRPCTransaction(tx *types.Transaction, blockHash common.Hash, blockNumber uint64, index uint64) *RPCTransaction {
	var signer types.Signer = types.FrontierSigner{}
	if tx.Type() == types.AccessListTxType {
		signer = types.NewEIP2718Signer(tx.ChainId())
	} else if tx.Protected() {
		signer = types.NewEIP155Signer(tx.ChainId())
	}
```
Like, I guess this is another of those places where we need to decide the signer based on the type: 
```
func (w *Wallet) SignTx(account accounts.Account, tx *types.Transaction, chainID *big.Int) (*types.Transaction, error) {
	signer := types.NewEIP155Signer(chainID)

```
And here: 
```
func (s *PublicTransactionPoolAPI) PendingTransactions() ([]*RPCTransaction, error) {
	pending, err := s.b.GetPoolTransactions()
	if err != nil {
		return nil, err
	}
	accounts := make(map[common.Address]struct{})
	for _, wallet := range s.b.AccountManager().Wallets() {
		for _, account := range wallet.Accounts() {
			accounts[account.Address] = struct{}{}
		}
	}
	transactions := make([]*RPCTransaction, 0, len(pending))
	for _, tx := range pending {
		var signer types.Signer = types.HomesteadSigner{}
		if tx.Protected() {
			signer = types.NewEIP155Signer(tx.ChainId())
		}
```

And here? 
```
func (tx *Transaction) WithSignature(sig []byte, chainID *BigInt) (signedTx *Transaction, _ error) {
	var signer types.Signer = types.HomesteadSigner{}
	if chainID != nil {
		signer = types.NewEIP155Signer(chainID.bigint)
	}
	rawTx, err := tx.tx.WithSignature(signer, common.CopyBytes(sig))
	return &Transaction{rawTx}, err

```
I don't even know what this does, have never used these parts of the code: 
```
// NewKeyStoreTransactorWithChainID is a utility method to easily create a transaction signer from
// an decrypted key from a keystore.
func NewKeyStoreTransactorWithChainID(keystore *keystore.KeyStore, account accounts.Account, chainID *big.Int) (*TransactOpts, error) {
	if chainID == nil {
		return nil, ErrNoChainID
	}
	signer := types.NewEIP155Signer(chainID)
```

jochem-brouwer commented on core/types/block_test.go:
```diff
@@ -59,6 +59,71 @@ func TestBlockEncoding(t *testing.T) {
+
+	check("len(Transactions)", len(block.Transactions()), 2)
+	check("Transactions[0].Hash", block.Transactions()[0].Hash(), tx1.Hash())
+	check("Transactions[1].Hash", block.Transactions()[1].Hash(), tx2.Hash())
```
If the `transactionsRoot` is not checked, isn't this a no-op? Shouldn't there be a hardcoded hash here, or at least a check to see that the reported transaction trie root is correct?

lightclient replied:
This is just checking that the transactions within the block structure can be decoded correctly, which is slightly different than ensuring the `transactionRoot` is correct.

holiman commented on core/types/transaction_marshalling.go:
```diff
@@ -134,10 +125,10 @@ func (t *Transaction) UnmarshalJSON(input []byte) error {
-		if dec.AccessList == nil {
-			return errors.New("missing required field 'accessList' in transaction")
+		// Access list is optional for now.
+		if dec.AccessList != nil {
+			itx.Accesses = dec.AccessList
```
Not sure about this. Why do we allow a `nil` list on a accesslist-type transaction?

fjl replied:
This is needed because we have `,omitempty` on `Accesses`. If the transaction has an empty access list, it will not have "accessList" in JSON output, so we can't require it for JSON input.

holiman commented on core/state_processor.go:
```diff
@@ -122,7 +110,12 @@ func applyTransaction(msg types.Message, config *params.ChainConfig, bc ChainCon
 	// Create a new receipt for the transaction, storing the intermediate root and gas used by the tx
 	// based on the eip phase, we're passing whether the root touch-delete accounts.
-	receipt := types.NewReceipt(root, result.Failed(), *usedGas)
+	var receipt *types.Receipt
+	if config.IsYoloV3(header.Number) {
+		receipt = types.NewEIP2718Receipt(tx.Type(), root, result.Failed(), *usedGas)
+	} else {
+		receipt = types.NewReceipt(root, result.Failed(), *usedGas)
+	}
```
@fjl with the changes in how tx validation works now, shouldn't we be able to replace this entire thing with 
```
receipt = types.NewEIP2718Receipt(tx.Type(), root, result.Failed(), *usedGas)
```
Since the tx type is already verified earlier, and if it's included, we'll trust the type to be correct.

holiman commented on core/types/transaction.go:
```diff
@@ -337,7 +337,7 @@ func (tx *Transaction) WithSignature(signer Signer, sig []byte) (*Transaction, e
 	}
 	cpy := tx.inner.copy()
-	cpy.setSignatureValues(v, r, s)
+	cpy.setSignatureValues(signer.ChainID(), v, r, s)
```
Shouldn't this method also raise an error if `signer.ChainID` differs from `tx.ChainID` ?

fjl replied:
The check is done in SignatureValues, so the error will be raised from there.

holiman commented on core/types/transaction_signing.go:
```diff
@@ -194,14 +205,19 @@ func (s eip2930Signer) Sender(tx *Transaction) (common.Address, error) {
+	switch txdata := tx.inner.(type) {
+	case *LegacyTx:
 		R, S, V = decodeSignature(sig)
 		if s.chainId.Sign() != 0 {
```
What if the legacy `chainId` is `nil`, in the case of a homestead-signed tx. Don't we need to check for that?

fjl replied:
`s.chainId` is always non-nil.
